clang/lib/AST/ASTContext.cpp

```diff
@@ -10078,6 +10078,12 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
       }
     }
 
+    // Methods explcitly marked with 'sycl_device' attribute (via SYCL_EXTERNAL)
+    // must be emitted regardless of number of actual uses
+    if (LangOpts.SYCLIsDevice && isa<CXXMethodDecl>(D))
+      if (auto *A = D->getAttr<SYCLDeviceAttr>())
+        return !A->isImplicit();
```

Comment:
Not sure if it worth to check for explicit `SYCLDeviceAttr`. Also, it would be good to hear feedback about this `if` in general: is it the right place and proper `if`?

Reply:
I'm ok with the place. I believe that in a happy future we will outline device code in other way: do this in CodeGen like OpenMP. We already try to do this in llorg, and we do some changes in this function for this. 
So, could you please make your changes to this function not conflicting with https://reviews.llvm.org/D71016#change-cNmcQq4FKHx8 ?
And, I'm not sure that checking for explicit `SYCLDeviceAttr` is needed.